File: photobooth.py
import asyncio
import os
import base64
import uuid
import sounddevice as sd
import numpy as np
import boto3
import qrcode
import cv2
from google import genai
from google.genai.live import AsyncSession
from google.genai.types import (
    FunctionDeclaration,
    FunctionResponse,
    GoogleSearch,
    Tool,
    LiveConnectConfig,
    Modality,
    ProactivityConfig,
    SpeechConfig,
    VoiceConfig,
    PrebuiltVoiceConfig,
    Schema,
    Type,
)

# Configure Gemini
api_key = os.getenv("GOOGLE_API_KEY")
print(api_key)
client = genai.Client(api_key=api_key)
S3_BUCKET = os.getenv("BUCKET_NAME", "festival-booth")

SAMPLERATE = 16000  # Rata de eșantionare pe care o așteaptă Gemini
CHANNELS = 1  # Mono
DTYPE = "int16"  # Formatul de date așteptat de API


# 🎫 Mock GraphQL
async def identify_user(qr: str) -> dict:
    return {"name": "Adrian", "code": qr}


# 📸 Funcții callback
def capture_snapshot() -> str:
    cap = cv2.VideoCapture(0)
    ret, frame = cap.read()
    cap.release()
    _, buf = cv2.imencode(".jpg", frame)
    return base64.b64encode(buf).decode()


def upload_to_s3(bytes_b64: str, user_code: str) -> str:
    s3 = boto3.client("s3")
    data = base64.b64decode(bytes_b64)
    key = f"{user_code}/{uuid.uuid4()}.jpg"
    s3.put_object(Bucket=S3_BUCKET, Key=key, Body=data, ACL="public-read")
    return f"https://{S3_BUCKET}.s3.amazonaws.com/{key}"


def generate_qr(url: str) -> str:
    img = qrcode.make(url)
    buf = cv2.imencode(".png", cv2.cvtColor(np.array(img), cv2.COLOR_RGB2BGR))[1]
    return base64.b64encode(buf).decode()


def display_on_tv(img_b64: str):
    data = base64.b64decode(img_b64)
    img = cv2.imdecode(np.frombuffer(data, np.uint8), cv2.IMREAD_COLOR)
    cv2.namedWindow("TV", cv2.WND_PROP_FULLSCREEN)
    cv2.setWindowProperty("TV", cv2.WND_PROP_FULLSCREEN, cv2.WINDOW_FULLSCREEN)
    cv2.imshow("TV", img)
    cv2.waitKey(60000)
    cv2.destroyAllWindows()


async def inputstream_generator():
    """Generator care produce blocuri de date de la microfon."""
    q_in = asyncio.Queue()
    loop = asyncio.get_event_loop()

    def callback(indata, frame_count, time_info, status):
        """Callback-ul este apelat de sounddevice într-un thread separat."""
        if status:
            print(status, flush=True)
        # Pune datele în coada asyncio într-un mod thread-safe
        loop.call_soon_threadsafe(q_in.put_nowait, (indata.copy(), status))

    # Folosim configurația centralizată
    stream = sd.InputStream(
        callback=callback, channels=CHANNELS, samplerate=SAMPLERATE, dtype=DTYPE
    )
    with stream:
        print("🎙️ Microfonul este activ...")
        while True:
            indata, status = await q_in.get()
            yield indata, status


async def send_audio_to_agent(session, **kwargs):
    """Show minimum and maximum value of each incoming audio block."""
    async for indata, status in inputstream_generator(**kwargs):
        if status:
            print(status)
        await session.send_realtime_input(
            audio={
                "data": indata.tobytes(),
                "mime_type": f"audio/pcm;rate={SAMPLERATE}",
            }
        )


def start_audio_output():
    out = sd.OutputStream(channels=1, samplerate=24000, dtype="int16")
    out.start()
    return out


async def handle_agent_messages(
    session: AsyncSession, audio_out: sd.OutputStream, user: dict
):
    while True:
        async for msg in session.receive():
            if msg.data:
                pcm = np.frombuffer(msg.data, dtype=np.int16)
                audio_out.write(pcm)
            if msg.tool_call and msg.tool_call.function_calls is not None:
                print("tool calls received")
                fn_responses: list[FunctionResponse] = []
                for fn_call in msg.tool_call.function_calls:
                    name = fn_call.name
                    args = fn_call.args or {}
                    if name == "capture_snapshot":
                        res = capture_snapshot()
                    elif name == "upload_to_s3":
                        res = upload_to_s3(args["bytes"], user["code"])
                    elif name == "generate_qr":
                        res = generate_qr(args["url"])
                    elif name == "display_on_tv":
                        res = display_on_tv(args["img_b64"]) or "ok"
                    else:
                        continue
                    fn_responses.append(
                        FunctionResponse(name=name, response={"output": res})
                    )
                await session.send_tool_response(function_responses=fn_responses)


# ✅ Flux principal
async def main():
    cap = cv2.VideoCapture(0)
    detector = cv2.QRCodeDetector()
    qr = None
    print("📸 Scanează QR de pe bilet...")
    while qr is None:
        ret, frame = cap.read()
        if not ret:
            continue
        cv2.imshow("Scanare QR", frame)
        data, _, _ = detector.detectAndDecode(frame)
        if data:
            qr = data
            break
        if cv2.waitKey(1) & 0xFF == ord("q"):
            return
    cap.release()
    cv2.destroyAllWindows()

    user = await identify_user(qr)
    print("🎟️ Welcome:", user)

    tools = [
        Tool(
            google_search=GoogleSearch(),
            function_declarations=[
                FunctionDeclaration(
                    name="capture_snapshot",
                    description="Take webcam photo",
                    parameters=Schema(type=Type.OBJECT, properties={}, required=[]),
                ),
                FunctionDeclaration(
                    name="upload_to_s3",
                    description="Upload photo",
                    parameters=Schema(
                        type=Type.OBJECT,
                        properties={
                            "bytes": Schema(type=Type.STRING),
                        },
                        required=["bytes"],
                    ),
                ),
                FunctionDeclaration(
                    name="generate_qr",
                    description="Make QR code",
                    parameters=Schema(
                        type=Type.OBJECT,
                        properties={
                            "url": Schema(type=Type.STRING),
                        },
                        required=["url"],
                    ),
                ),
                FunctionDeclaration(
                    name="display_on_tv",
                    description="Display image on TV",
                    parameters=Schema(
                        type=Type.OBJECT,
                        properties={"img_b64": Schema(type=Type.STRING)},
                        required=["img_b64"],
                    ),
                ),
            ]
        )
    ]

    async with client.aio.live.connect(
        model="gemini-2.5-flash-preview-native-audio-dialog",
        # model="gemini-2.0-flash-live-001",
        config=LiveConnectConfig(
            response_modalities=[Modality.AUDIO],
            system_instruction=f"Esti un agent de photobooth la festivalul diffusion (pronuntat ca in engleza) si vorbesti cu {user.get('name')}",
            speech_config=SpeechConfig(
                voice_config=VoiceConfig(
                    prebuilt_voice_config=PrebuiltVoiceConfig(voice_name="Zephyr")
                )
            ),
            tools=tools,
        ),
    ) as session:
        print("📢 Starting session...")
        audio_out = start_audio_output()

        async with asyncio.TaskGroup() as tg:
            mic = tg.create_task(send_audio_to_agent(session=session))
            await tg.create_task(
                handle_agent_messages(session=session, audio_out=audio_out, user=user)
            )


if __name__ == "__main__":
    asyncio.run(main())

File: photobooth.go
// main.go
package main

import (
    "context"
    "encoding/base64"
    "bytes"
    "fmt"
    "log"
    "os"
    "os/signal"
    "time"

    "gocv.io/x/gocv"
    "github.com/google/uuid"
    "github.com/gordonklaus/portaudio"
    "layeh.com/gopus"
    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/s3"
    "github.com/skip2/go-qrcode"
    "google.golang.org/genai"
    "google.golang.org/genai/genaiws"
)

// Canale audio/video
var (
    audioChan = make(chan []byte, 100)
    videoChan = make(chan []byte, 10)
)

func main() {
    stopScanner := make(chan struct{})
    qrChan := make(chan string)

    go webcamQRScanner(videoChan, stopScanner, qrChan)
    qrData := <-qrChan
    close(stopScanner)

    log.Println("QR detectat:", qrData)
    user, err := graphqlIdentify(qrData)
    if err != nil {
        log.Fatalf("Grafică utilizator eșuat: %v", err)
    }
    log.Printf("User: %s (%s)", user.Name, user.Code)

    go streamMicAudio(audioChan)
    go playAudio(audioChan)
    stopVid := make(chan struct{})
    go webcamStream(videoChan, stopVid)

    startGeminiLive(user)

    close(stopVid)
    log.Println("Sesiune terminată")
}

func startGeminiLive(user *User) {
    ctx := context.Background()
    client, err := genai.NewClient(ctx, &genai.ClientConfig{APIKey: os.Getenv("GEMINI_API_KEY")})
    if err != nil {
        log.Fatal(err)
    }

    tools := []genaiws.FunctionDeclaration{
        genaiws.FunctionDeclaration{Name: "captureSnapshot", Description: "Makes snapshot", Parameters: schema(`{"type":"object","properties":{},"required":[]}`)},
        genaiws.FunctionDeclaration{Name: "uploadToS3", Description: "Uploads image", Parameters: schema(`{"type":"object","properties":{"bytes":{"type":"string"},"user_code":{"type":"string"}},"required":["bytes","user_code"]}`)},
        genaiws.FunctionDeclaration{Name: "generateQR", Description: "Creates QR code", Parameters: schema(`{"type":"object","properties":{"url":{"type":"string"}},"required":["url"]}`)},
        genaiws.FunctionDeclaration{Name: "displayOnTV", Description: "Show on TV", Parameters: schema(`{"type":"object","properties":{"img_b64":{"type":"string"}},"required":["img_b64"]}`)},
    }

    conn, err := genaiws.DialLive(ctx, client, "gemini-2.0-flash-live-preview-04-09", genaiws.LiveConfig{
        ResponseModalities: []genaiws.Modality{genaiws.ModalityAudio, genaiws.ModalityText},
        Tools:              tools,
    })
    if err != nil {
        log.Fatal(err)
    }
    defer conn.Close()

    systemPrompt := fmt.Sprintf("Ești un asistent foto la festival. Salută pe %s și propune o poză.", user.Name)
    conn.SendText(systemPrompt)

    go func() {
        for msg := range conn.Receive() {
            if fc := msg.FunctionCall; fc != nil {
                switch fc.Name {
                case "captureSnapshot":
                    img := <-videoChan
                    conn.SendToolResponse("captureSnapshot", map[string]string{"image_bytes": base64.StdEncoding.EncodeToString(img)})
                case "uploadToS3":
                    b, _ := base64.StdEncoding.DecodeString(fc.Args["bytes"].(string))
                    url, _ := uploadToS3(user.Code, b)
                    conn.SendToolResponse("uploadToS3", map[string]string{"url": url})
                case "generateQR":
                    png, _ := qrcode.Encode(fc.Args["url"].(string), qrcode.Medium, 256)
                    conn.SendToolResponse("generateQR", map[string]string{"qr_b64": base64.StdEncoding.EncodeToString(png)})
                case "displayOnTV":
                    displayOnTV(fc.Args["img_b64"].(string))
                    conn.SendToolResponse("displayOnTV", map[string]string{"status": "ok"})
                }
            }
        }
    }()

    c := make(chan os.Signal, 1)
    signal.Notify(c, os.Interrupt)
    <-c
}

type User struct{ Name, Code string }

func graphqlIdentify(qr string) (*User, error) {
    // Înlocuiește cu apel real GraphQL
    return &User{Name: "Adrian", Code: "WR1234"}, nil
}

func streamMicAudio(out chan<- []byte) {
    portaudio.Initialize(); defer portaudio.Terminate()
    enc, _ := gopus.NewEncoder(48000, 1, gopus.AppAudio)
    stream, _ := portaudio.OpenDefaultStream(1, 0, 48000, 960, func(in []int16) {
        pcm := make([]float32, len(in))
        for i := range in {
            pcm[i] = float32(in[i]) / 32768
        }
        if data, err := enc.Encode(pcm, 960, 4000); err == nil {
            out <- data
        }
    })
    defer stream.Close()
    stream.Start()
    select {}
}

func playAudio(in <-chan []byte) {
    portaudio.Initialize(); defer portaudio.Terminate()
    dec, _ := gopus.NewDecoder(48000, 1)
    stream, _ := portaudio.OpenDefaultStream(0, 1, 48000, 960, func(out []int16) {
        select {
        case data := <-in:
            pcm := make([]int16, 960)
            if _, err := dec.Decode(data, pcm, false); err == nil {
                copy(out, pcm)
            }
        default:
            for i := range out {
                out[i] = 0
            }
        }
    })
    defer stream.Close()
    stream.Start()
    select {}
}

func webcamStream(out chan<- []byte, stop <-chan struct{}) {
    cam, err := gocv.VideoCaptureDevice(0)
    if err != nil { log.Fatalf("webcam error: %v", err) }
    defer cam.Close()
    img := gocv.NewMat(); defer img.Close()
    ticker := time.NewTicker(100 * time.Millisecond)
    defer ticker.Stop()

    for {
        select {
        case <-ticker.C:
            if ok := cam.Read(&img); !ok || img.Empty() {
                continue
            }
            if buf, err := gocv.IMEncode(".jpg", img); err == nil {
                out <- buf.GetBytes()
            }
        case <-stop:
            return
        }
    }
}

func webcamQRScanner(out chan<- []byte, stop <-chan struct{}, qrChan chan<- string) {
    cam, _ := gocv.VideoCaptureDevice(0)
    defer cam.Close()
    img := gocv.NewMat(); defer img.Close()
    qcd := gocv.NewQRCodeDetector()
    defer qcd.Close()
    ticker := time.NewTicker(100 * time.Millisecond)
    defer ticker.Stop()

    for {
        select {
        case <-ticker.C:
            if ok := cam.Read(&img); !ok || img.Empty() {
                continue
            }
            if buf, err := gocv.IMEncode(".jpg", img); err == nil {
                out <- buf.GetBytes()
            }
            if code := qcd.DetectAndDecode(img); code != "" {
                qrChan <- code
                return
            }
        case <-stop:
            return
        }
    }
}

func uploadToS3(userCode string, data []byte) (string, error) {
    sess := session.Must(session.NewSession())
    svc := s3.New(sess)
    key := fmt.Sprintf("%s/%s.jpg", userCode, uuid.New().String())
    _, err := svc.PutObject(&s3.PutObjectInput{
        Bucket: aws.String("festival-booth"), Key: aws.String(key), Body: bytes.NewReader(data),
    })
    return fmt.Sprintf("https://festival-booth.s3.amazonaws.com/%s", key), err
}

func displayOnTV(imgB64 string) {
    data, err := base64.StdEncoding.DecodeString(imgB64)
    if err != nil { log.Printf("TV decode err: %v", err); return }
    mat, err := gocv.IMDecode(data, gocv.IMReadColor)
    if err != nil { log.Printf("IMDecode err: %v", err); return }
    defer mat.Close()
    win := gocv.NewWindow("FestivalBooth")
    defer win.Close()
    win.SetWindowProperty(gocv.WindowPropertyFullscreen, gocv.WindowFullscreen)
    win.IMShow(mat)
    win.WaitKey(60000)
}

func schema(s string) interface{} { return map[string]interface{}{} }
